NPE for http sink when restore from checkpoint

### Description

I use this repo as **sink connector** for flink sql, but I found that the job can't restore from checkpoint and throw NullPointerException when the job crash. Except this case, I just start the flink app from checkpoint, it also throw NPE.

### Used Version
flink version 1.16.3
flink-http-connector 0.16.0

### Expected Behavior
The flink sql job can restore from checkpoint

### Flink SQL Example
My example like below:

```java
    // http sink
    public static String httpSinkDDL = "CREATE TEMPORARY TABLE http_sink_table (\n" +
            "  `region` STRING,\n" +
            "  `region_uv` BIGINT\n" +
            ") WITH (\n" +
            "  'connector' = 'http-sink',\n" +
            "  'url' = 'url" +
            "  'insert-method' = 'POST',\n" +
            "  'gid.connector.http.sink.header.Content-Type' = 'application/json',\n" +
            "  'gid.connector.http.sink.writer.request.mode' = 'batch',\n" +
            "  'gid.connector.http.sink.request.batch.size' = '100',\n" +
            "  'format' = 'json'\n" +
            ");";

    public static String transformDMLPT = "INSERT INTO http_sink_table \n" +
            "SELECT\n" +
            "  region,\n" +
            "  COUNT(DISTINCT account_id) AS region_uv\n" +
            "FROM (\n" +
            "  SELECT\n" +
            "    s.account_id,\n" +
            "    s.proc_time,\n" +
            "    COALESCE(LangToRegionFuc(h.f.LANG), 'en') AS region \n" +
            "  FROM (\n" +
            "    SELECT\n" +
            "      CONCAT(SUBSTRING(MD5(t2.`#account_id`), 1, 4), '|', t2.`#account_id`) AS account_id,\n" +
            "      source.proc_time AS proc_time \n" +
            "    FROM\n" +
            "      kafka_source_table AS source,\n" +
            "      UNNEST(source.data_object.data) AS t2 (`#account_id`, `#event_name`, `properties`)\n" +
            "    WHERE\n" +
            "      t2.`#event_name` = 'Match' and t2.properties.action = 'join_bellmatch' \n" +
            "  ) AS s\n" +
            "  JOIN hbase_region_table FOR SYSTEM_TIME AS OF s.proc_time AS h\n" +
            "  ON s.account_id = h.rk\n" +
            ")\n" +
            "GROUP BY\n" +
            "  HOP(proc_time, INTERVAL '1' MINUTE, INTERVAL '3' HOUR),\n" +
            "  region;";
```

### Exception Logs

```text
2025-11-24 19:58:20,954 INFO  org.apache.flink.runtime.taskmanager.Task                    [] - Attempting to fail task externally GroupWindowAggregate[8] -> http_sink_table[9]: Writer (2/2)#5 (a93410a57893a814791e5f83ced20de0_5e4132aa89d205854be108b7fd747a9c_1_5).
2025-11-24 19:58:20,955 WARN  org.apache.flink.runtime.taskmanager.Task                    [] - GroupWindowAggregate[8] -> http_sink_table[9]: Writer (2/2)#5 (a93410a57893a814791e5f83ced20de0_5e4132aa89d205854be108b7fd747a9c_1_5) switched from INITIALIZING to FAILED with failure cause: org.apache.flink.runtime.taskmanager.AsynchronousException: Caught exception while processing timer.
	at org.apache.flink.streaming.runtime.tasks.StreamTask$StreamTaskAsyncExceptionHandler.handleAsyncException(StreamTask.java:1573)
	at org.apache.flink.streaming.runtime.tasks.StreamTask.handleAsyncException(StreamTask.java:1548)
	at org.apache.flink.streaming.runtime.tasks.StreamTask.invokeProcessingTimeCallback(StreamTask.java:1688)
	at org.apache.flink.streaming.runtime.tasks.StreamTask.lambda$null$22(StreamTask.java:1677)
	at org.apache.flink.streaming.runtime.tasks.StreamTaskActionExecutor$1.runThrowing(StreamTaskActionExecutor.java:50)
	at org.apache.flink.streaming.runtime.tasks.mailbox.Mail.run(Mail.java:90)
	at org.apache.flink.streaming.runtime.tasks.mailbox.MailboxProcessor.runMail(MailboxProcessor.java:398)
	at org.apache.flink.streaming.runtime.tasks.mailbox.MailboxProcessor.processMailsNonBlocking(MailboxProcessor.java:383)
	at org.apache.flink.streaming.runtime.tasks.mailbox.MailboxProcessor.processMail(MailboxProcessor.java:345)
	at org.apache.flink.streaming.runtime.tasks.mailbox.MailboxProcessor.runMailboxLoop(MailboxProcessor.java:229)
	at org.apache.flink.streaming.runtime.tasks.StreamTask.restoreInternal(StreamTask.java:709)
	at org.apache.flink.streaming.runtime.tasks.StreamTask.restore(StreamTask.java:672)
	at org.apache.flink.runtime.taskmanager.Task.runWithSystemExitMonitoring(Task.java:935)
	at org.apache.flink.runtime.taskmanager.Task.restoreAndInvoke(Task.java:904)
	at org.apache.flink.runtime.taskmanager.Task.doRun(Task.java:728)
	at org.apache.flink.runtime.taskmanager.Task.run(Task.java:550)
	at java.base/java.lang.Thread.run(Unknown Source)
Caused by: TimerException{org.apache.flink.streaming.runtime.tasks.ExceptionInChainedOperatorException: Could not forward element to next operator}
	... 15 more
Caused by: org.apache.flink.streaming.runtime.tasks.ExceptionInChainedOperatorException: Could not forward element to next operator
	at org.apache.flink.streaming.runtime.tasks.CopyingChainingOutput.pushToOperator(CopyingChainingOutput.java:99)
	at org.apache.flink.streaming.runtime.tasks.CopyingChainingOutput.collect(CopyingChainingOutput.java:57)
	at org.apache.flink.streaming.runtime.tasks.CopyingChainingOutput.collect(CopyingChainingOutput.java:29)
	at org.apache.flink.streaming.api.operators.CountingOutput.collect(CountingOutput.java:56)
	at org.apache.flink.streaming.api.operators.CountingOutput.collect(CountingOutput.java:29)
	at org.apache.flink.streaming.api.operators.TimestampedCollector.collect(TimestampedCollector.java:51)
	at org.apache.flink.table.runtime.operators.window.AggregateWindowOperator.collect(AggregateWindowOperator.java:218)
	at org.apache.flink.table.runtime.operators.window.AggregateWindowOperator.emitWindowResult(AggregateWindowOperator.java:208)
	at org.apache.flink.table.runtime.operators.window.WindowOperator.onProcessingTime(WindowOperator.java:405)
	at org.apache.flink.streaming.api.operators.InternalTimerServiceImpl.onProcessingTime(InternalTimerServiceImpl.java:284)
	at org.apache.flink.streaming.runtime.tasks.StreamTask.invokeProcessingTimeCallback(StreamTask.java:1686)
	... 14 more
Caused by: java.lang.NullPointerException
	at org.apache.flink.formats.json.JsonRowDataSerializationSchema.serialize(JsonRowDataSerializationSchema.java:99)
	at org.apache.flink.formats.json.JsonRowDataSerializationSchema.serialize(JsonRowDataSerializationSchema.java:43)
	at com.getindata.connectors.http.internal.table.SerializationSchemaElementConverter.apply(SerializationSchemaElementConverter.java:51)
	at com.getindata.connectors.http.internal.table.SerializationSchemaElementConverter.apply(SerializationSchemaElementConverter.java:14)
	at org.apache.flink.connector.base.sink.writer.AsyncSinkWriter.write(AsyncSinkWriter.java:327)
	at org.apache.flink.streaming.runtime.operators.sink.SinkWriterOperator.processElement(SinkWriterOperator.java:160)
	at org.apache.flink.streaming.runtime.tasks.CopyingChainingOutput.pushToOperator(CopyingChainingOutput.java:82)
	... 24 more
```

###Reproduce

Actually this issue can always can be reproduced as long as the job start from checkpoint.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,6 +1,8 @@
 # Changelog
 
 ## [Unreleased]
+### Fixed
+- Fixed http sink NEP issue when flink job restore from checkpoint.
 
 ## [0.23.0] - 2025-11-07
 

--- a/src/main/java/com/getindata/connectors/http/internal/sink/HttpSinkInternal.java
+++ b/src/main/java/com/getindata/connectors/http/internal/sink/HttpSinkInternal.java
@@ -150,6 +150,11 @@ public class HttpSinkInternal<InputT> extends AsyncSinkBase<InputT, HttpSinkRequ
                 Collection<BufferedRequestState<HttpSinkRequestEntry>> recoveredState)
             throws IOException {
 
+        ElementConverter<InputT, HttpSinkRequestEntry> elementConverter = getElementConverter();
+        if (elementConverter instanceof SchemaLifecycleAwareElementConverter) {
+            ((SchemaLifecycleAwareElementConverter<?, ?>) elementConverter).open(context);
+        }
+
         return new HttpSinkWriter<>(
             getElementConverter(),
             context,

--- a/src/main/java/com/getindata/connectors/http/internal/sink/HttpSinkInternal.java
+++ b/src/main/java/com/getindata/connectors/http/internal/sink/HttpSinkInternal.java
@@ -143,7 +143,7 @@ public class HttpSinkInternal<InputT> extends AsyncSinkBase<InputT, HttpSinkRequ
     private ElementConverter<InputT, HttpSinkRequestEntry> initElementConverterOfSchema(InitContext context) {
         ElementConverter<InputT, HttpSinkRequestEntry> elementConverter = getElementConverter();
         if (elementConverter instanceof SchemaLifecycleAwareElementConverter) {
-            ((SchemaLifecycleAwareElementConverter<?, ?>) elementConverter).open(context);
+            elementConverter.open(context);
         }
         return elementConverter;
     }

--- a/src/main/java/com/getindata/connectors/http/internal/sink/HttpSinkInternal.java
+++ b/src/main/java/com/getindata/connectors/http/internal/sink/HttpSinkInternal.java
@@ -117,11 +117,7 @@ public class HttpSinkInternal<InputT> extends AsyncSinkBase<InputT, HttpSinkRequ
     public StatefulSinkWriter<InputT, BufferedRequestState<HttpSinkRequestEntry>> createWriter(
             InitContext context) throws IOException {
 
-        ElementConverter<InputT, HttpSinkRequestEntry> elementConverter = getElementConverter();
-        if (elementConverter instanceof SchemaLifecycleAwareElementConverter) {
-            // This cast is needed for Flink 1.15.3 build
-            ((SchemaLifecycleAwareElementConverter<?, ?>) elementConverter).open(context);
-        }
+        ElementConverter<InputT, HttpSinkRequestEntry> elementConverter = initElementConverterOfSchema(context);
 
         return new HttpSinkWriter<>(
             elementConverter,
@@ -144,16 +140,21 @@ public class HttpSinkInternal<InputT> extends AsyncSinkBase<InputT, HttpSinkRequ
         );
     }
 
+    private ElementConverter<InputT, HttpSinkRequestEntry> initElementConverterOfSchema(InitContext context) {
+        ElementConverter<InputT, HttpSinkRequestEntry> elementConverter = getElementConverter();
+        if (elementConverter instanceof SchemaLifecycleAwareElementConverter) {
+            ((SchemaLifecycleAwareElementConverter<?, ?>) elementConverter).open(context);
+        }
+        return elementConverter;
+    }
+
     @Override
     public StatefulSinkWriter<InputT, BufferedRequestState<HttpSinkRequestEntry>> restoreWriter(
                 InitContext context,
                 Collection<BufferedRequestState<HttpSinkRequestEntry>> recoveredState)
             throws IOException {
 
-        ElementConverter<InputT, HttpSinkRequestEntry> elementConverter = getElementConverter();
-        if (elementConverter instanceof SchemaLifecycleAwareElementConverter) {
-            ((SchemaLifecycleAwareElementConverter<?, ?>) elementConverter).open(context);
-        }
+        initElementConverterOfSchema(context);
 
         return new HttpSinkWriter<>(
             getElementConverter(),

--- a/src/test/java/com/getindata/connectors/http/internal/sink/HttpSinkInternalTest.java
+++ b/src/test/java/com/getindata/connectors/http/internal/sink/HttpSinkInternalTest.java
@@ -5,14 +5,13 @@ import com.getindata.connectors.http.SchemaLifecycleAwareElementConverter;
 import com.getindata.connectors.http.internal.HeaderPreprocessor;
 import com.getindata.connectors.http.internal.SinkHttpClient;
 import com.getindata.connectors.http.internal.SinkHttpClientBuilder;
+import java.util.Collections;
+import java.util.Properties;
 import org.apache.flink.api.connector.sink2.Sink;
 import org.apache.flink.connector.base.sink.writer.ElementConverter;
 import org.apache.flink.metrics.groups.OperatorIOMetricGroup;
 import org.apache.flink.metrics.groups.SinkWriterMetricGroup;
 import org.junit.jupiter.api.Test;
-
-import java.util.Collections;
-import java.util.Properties;
 
 import static org.mockito.Mockito.*;
 
@@ -27,7 +26,8 @@ public class HttpSinkInternalTest {
         OperatorIOMetricGroup mockIOMetricGroup = mock(OperatorIOMetricGroup.class);
         when(mockMetricGroup.getIOMetricGroup()).thenReturn(mockIOMetricGroup);
 
-        HttpSinkInternal<Object> httpSink = createTestSink((ElementConverter<Object, HttpSinkRequestEntry>) mockConverter);
+        HttpSinkInternal<Object> httpSink = createTestSink(
+                (ElementConverter<Object, HttpSinkRequestEntry>) mockConverter);
         httpSink.createWriter(mockContext);
         // com.getindata.connectors.http.internal.sink.HttpSinkInternal.initElementConverterOfSchema
         // org.apache.flink.connector.base.sink.writer.AsyncSinkWriter
@@ -42,7 +42,8 @@ public class HttpSinkInternalTest {
         SinkWriterMetricGroup mockMetricGroup = mock(SinkWriterMetricGroup.class);
         when(mockContext.metricGroup()).thenReturn(mockMetricGroup);
         when(mockMetricGroup.getIOMetricGroup()).thenReturn(mock(OperatorIOMetricGroup.class));
-        HttpSinkInternal<Object> httpSink = createTestSink((ElementConverter<Object, HttpSinkRequestEntry>) mockConverter);
+        HttpSinkInternal<Object> httpSink = createTestSink(
+                (ElementConverter<Object, HttpSinkRequestEntry>) mockConverter);
 
         httpSink.restoreWriter(mockContext, Collections.emptyList());
         // com.getindata.connectors.http.internal.sink.HttpSinkInternal.initElementConverterOfSchema
@@ -58,7 +59,8 @@ public class HttpSinkInternalTest {
 
         return new HttpSinkInternal<>(
                 converter,
-                10, 1, 20, 1024L, 1000L, 1024L,
+                10, 1, 20,
+                1024L, 1000L, 1024L,
                 "http://test-endpoint.com",
                 mock(HttpPostRequestCallback.class),
                 mock(HeaderPreprocessor.class),

--- a/src/test/java/com/getindata/connectors/http/internal/sink/HttpSinkInternalTest.java
+++ b/src/test/java/com/getindata/connectors/http/internal/sink/HttpSinkInternalTest.java
@@ -1,0 +1,70 @@
+package com.getindata.connectors.http.internal.sink;
+
+import com.getindata.connectors.http.HttpPostRequestCallback;
+import com.getindata.connectors.http.SchemaLifecycleAwareElementConverter;
+import com.getindata.connectors.http.internal.HeaderPreprocessor;
+import com.getindata.connectors.http.internal.SinkHttpClient;
+import com.getindata.connectors.http.internal.SinkHttpClientBuilder;
+import org.apache.flink.api.connector.sink2.Sink;
+import org.apache.flink.connector.base.sink.writer.ElementConverter;
+import org.apache.flink.metrics.groups.OperatorIOMetricGroup;
+import org.apache.flink.metrics.groups.SinkWriterMetricGroup;
+import org.junit.jupiter.api.Test;
+
+import java.util.Collections;
+import java.util.Properties;
+
+import static org.mockito.Mockito.*;
+
+public class HttpSinkInternalTest {
+
+    @Test
+    void testCreateWriterInitializesLifecycleAwareConverter() throws Exception {
+        SchemaLifecycleAwareElementConverter<?, ?> mockConverter = mock(SchemaLifecycleAwareElementConverter.class);
+        Sink.InitContext mockContext = mock(Sink.InitContext.class);
+        SinkWriterMetricGroup mockMetricGroup = mock(SinkWriterMetricGroup.class);
+        when(mockContext.metricGroup()).thenReturn(mockMetricGroup);
+        OperatorIOMetricGroup mockIOMetricGroup = mock(OperatorIOMetricGroup.class);
+        when(mockMetricGroup.getIOMetricGroup()).thenReturn(mockIOMetricGroup);
+
+        HttpSinkInternal<Object> httpSink = createTestSink((ElementConverter<Object, HttpSinkRequestEntry>) mockConverter);
+        httpSink.createWriter(mockContext);
+        // com.getindata.connectors.http.internal.sink.HttpSinkInternal.initElementConverterOfSchema
+        // org.apache.flink.connector.base.sink.writer.AsyncSinkWriter
+        verify(mockConverter, times(2)).open(mockContext);
+    }
+
+    @Test
+    void testRestoreWriterInitializesLifecycleAwareConverter() throws Exception {
+
+        SchemaLifecycleAwareElementConverter<?, ?> mockConverter = mock(SchemaLifecycleAwareElementConverter.class);
+        Sink.InitContext mockContext = mock(Sink.InitContext.class);
+        SinkWriterMetricGroup mockMetricGroup = mock(SinkWriterMetricGroup.class);
+        when(mockContext.metricGroup()).thenReturn(mockMetricGroup);
+        when(mockMetricGroup.getIOMetricGroup()).thenReturn(mock(OperatorIOMetricGroup.class));
+        HttpSinkInternal<Object> httpSink = createTestSink((ElementConverter<Object, HttpSinkRequestEntry>) mockConverter);
+
+        httpSink.restoreWriter(mockContext, Collections.emptyList());
+        // com.getindata.connectors.http.internal.sink.HttpSinkInternal.initElementConverterOfSchema
+        // org.apache.flink.connector.base.sink.writer.AsyncSinkWriter
+        verify(mockConverter, times(2)).open(mockContext);
+    }
+
+
+    private HttpSinkInternal<Object> createTestSink(ElementConverter<Object, HttpSinkRequestEntry> converter) {
+        SinkHttpClientBuilder mockClientBuilder = mock(SinkHttpClientBuilder.class);
+        when(mockClientBuilder.build(any(), any(), any(), any()))
+                .thenReturn(mock(SinkHttpClient.class));
+
+        return new HttpSinkInternal<>(
+                converter,
+                10, 1, 20, 1024L, 1000L, 1024L,
+                "http://test-endpoint.com",
+                mock(HttpPostRequestCallback.class),
+                mock(HeaderPreprocessor.class),
+                mockClientBuilder,
+                new Properties()
+        );
+    }
+
+}

--- a/src/test/java/com/getindata/connectors/http/internal/sink/HttpSinkInternalTest.java
+++ b/src/test/java/com/getindata/connectors/http/internal/sink/HttpSinkInternalTest.java
@@ -1,19 +1,24 @@
 package com.getindata.connectors.http.internal.sink;
 
-import com.getindata.connectors.http.HttpPostRequestCallback;
-import com.getindata.connectors.http.SchemaLifecycleAwareElementConverter;
-import com.getindata.connectors.http.internal.HeaderPreprocessor;
-import com.getindata.connectors.http.internal.SinkHttpClient;
-import com.getindata.connectors.http.internal.SinkHttpClientBuilder;
 import java.util.Collections;
 import java.util.Properties;
+
 import org.apache.flink.api.connector.sink2.Sink;
 import org.apache.flink.connector.base.sink.writer.ElementConverter;
 import org.apache.flink.metrics.groups.OperatorIOMetricGroup;
 import org.apache.flink.metrics.groups.SinkWriterMetricGroup;
 import org.junit.jupiter.api.Test;
+import static org.mockito.Mockito.any;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
 
-import static org.mockito.Mockito.*;
+import com.getindata.connectors.http.HttpPostRequestCallback;
+import com.getindata.connectors.http.SchemaLifecycleAwareElementConverter;
+import com.getindata.connectors.http.internal.HeaderPreprocessor;
+import com.getindata.connectors.http.internal.SinkHttpClient;
+import com.getindata.connectors.http.internal.SinkHttpClientBuilder;
 
 public class HttpSinkInternalTest {
 

